### PR TITLE
executor: support Intersection of IndexMerge

### LINKED ISSUE #39334: Support IndexMerge Intersection (executor part)
## Feature Request

sub task of https://github.com/pingcap/tidb/issues/39333

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #39334

Problem Summary: Support intersection of IndexMerge

### What is changed and how it works?
1. index_merge_reader.go: use a map to record handle count of each partial path. For partition table, we use one goroutine for one or multiple partitions.
2. mem_reader.go: Handle data in mem_table. Same logic as above, but only one goroutine, because the data in mem_table is not large.
### Check List

Tests 

- [ ] Unit test
- [x] Integration test: https://github.com/pingcap/tidb-test/pull/2021 and https://github.com/pingcap/tidb/pull/39323
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
executor: support Intersection of IndexMerge
```

### CODE REVIEW

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -118,6 +119,17 @@ type IndexMergeReaderExecutor struct {
...
+
+	// Whether it's intersection or union.
+	isIntersection bool
+}
+
+type indexMergeTableTask struct {
+	lookupTableTask
+
+	// workerID and parTblIdx are only used in indexMergeProcessWorker.fetchLoopIntersection.
+	workerID  int
```
Comment (time-and-fate) [inline executor/index_merge_reader.go:131]: Seems `workerID` is not used.

Comment (guo-shaoge) [reply executor/index_merge_reader.go]: Done

File: executor/index_merge_reader.go [hunk 2]
```diff
@@ -775,7 +789,157 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+	}
+	failpoint.Inject("testIndexMergeIntersectionConcurrency", func(val failpoint.Value) {
+		con := val.(int)
+		if con != workerCnt {
+			panic(fmt.Sprintf("unexpected workerCnt, expect %d, got %d", con, workerCnt))
+		}
+	})
+	handleMaps := make([]*kv.MemAwareHandleMap[*int], 0, partCnt)
+	workers := make([]*intersectionProcessWorker, 0, workerCnt)
+	wg := sync.WaitGroup{}
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:870]: We can use `util.WaitGroupWrapper`

Comment (wshwsh12) [reply executor/index_merge_reader.go:870]: And we need a recover function for the goroutine

**Review by time-and-fate — COMMENTED** [review]
I think we can have more detailed comments, like how we use the handle map to implement the intersection logic, how the concurrency is related to the dynamic pruning, etc.
And please fix the panic issue. Others LGTM.

File: executor/index_merge_reader.go (hunk 2 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go:896]: Done

Comment (wshwsh12) [reply executor/index_merge_reader.go:899]: We can use 
```
wg := util.WaitGroupWrapper{}
wg.RunWithRecover()
wg.Wait()
```

**Review by time-and-fate — APPROVED** [review]

File: executor/index_merge_reader.go [hunk 3]
```diff
@@ -775,18 +789,235 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+	}
+
+	// One goroutine may handle one or multiple partitions.
+	// Max number of partition number is 8192, we use ExecutorConcurrency to avoid too many goroutines.
+	maxWorkerCnt := w.indexMerge.ctx.GetSessionVars().IndexMergeIntersectionConcurrency()
+	partCnt := 1
+	workerCnt := 1
+	partCntPerWorker := 1
+
+	const maxChannelSize int = 1024
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: done

File: executor/index_merge_reader.go (hunk 2 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: done

File: executor/index_merge_reader.go [hunk 4]
```diff
@@ -775,18 +789,235 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+		go func() {
+			defer trace.StartRegion(ctx, "IndexMergeIntersectionProcessWorker").End()
+			defer wg.Done()
+			util.WithRecovery(
+				func() {
+					worker.doIntersectionPerPartition()
+				},
+				w.handleLoopFetcherPanic(ctx, resultCh, "IndexMergeIntersectionProcessWorker", errCh),
+			)
+		}()
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: Use `WaitGroupWrapper` instead

File: executor/index_merge_reader.go [hunk 5]
```diff
@@ -775,18 +789,217 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+	wg.Wait()
+	if processWorkerErr {
+		return
+	}
+
+	var totalMemUsage int64
+	for _, processWorker := range workers {
+		totalMemUsage += processWorker.totalMemUsage
+	}
+	w.indexMerge.memTracker.Consume(-(totalMemUsage))
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:979]: We can create a tracker for each w.memTrack,  use w.memTracker.Detach() to do this and remove the var totalMemUsage

File: executor/index_merge_reader.go [hunk 6]
```diff
@@ -775,18 +789,217 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+			} else {
+				cnt := 1
+				mapDelta += hMap.Set(h, &cnt) + int64(h.ExtraMemSize())
+				rowDelta += 1
+			}
+		}
+
+		logutil.BgLogger().Debug("intersectionProcessWorker handle tasks", zap.Int("workerID", w.workerID),
+			zap.Int("task.handles", len(task.handles)), zap.Int64("rowDelta", rowDelta))
+		if rowDelta > 0 {
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:845]: Don't need check rowDelta

File: executor/index_merge_reader.go [hunk 7]
```diff
@@ -775,18 +789,217 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+			w.rowDelta += rowDelta
+			w.memUsageDelta += mapDelta + (rowDelta * int64(unsafe.Sizeof(reflect.Int64)))
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:847]: Can add `(rowDelta * int64(unsafe.Sizeof(reflect.Int64)))` when `w.rowDelta >= int64(batchSize)`

Comment (wshwsh12) [reply executor/index_merge_reader.go:847]: And why use reflect.Int64? Use unsafe.Sizeof(int(0))?

File: executor/index_merge_reader.go [hunk 8]
```diff
@@ -775,18 +789,213 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+			break
+		}
+	}
+	for _, processWorker := range workers {
+		close(processWorker.workerCh)
+	}
+	wg.Wait()
+	if processWorkerErr {
+		return
+	}
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:973]: What is the check meaning?

Comment (guo-shaoge) [reply executor/index_merge_reader.go]: D

File: executor/index_merge_reader.go (hunk 7 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: D

File: executor/index_merge_reader.go (hunk 6 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: D

File: executor/index_merge_reader.go (hunk 5 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: D

File: executor/index_merge_reader.go [hunk 9]
```diff
@@ -775,18 +788,199 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
...
+	maxWorkerCnt := w.indexMerge.ctx.GetSessionVars().IndexMergeIntersectionConcurrency()
+	maxChannelSize := atomic.LoadInt32(&LookupTableTaskChannelSize)
+	batchSize := w.indexMerge.ctx.GetSessionVars().IndexLookupSize
+
+	partCnt := 1
+	if w.indexMerge.partitionTableMode {
+		partCnt = len(w.indexMerge.prunedPartitions)
+	}
+	workerCnt := partCnt
+	if workerCnt > maxWorkerCnt {
```
Comment (XuHuaiyu) [inline executor/index_merge_reader.go:924]: Use mathutil.Max is simpler

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**